File: 01_introduction_to_algorithms/typescript/binary-search.ts
function binarySearch(searchList: number[], value: number): number | null {
  let low = 0;
  let high = searchList.length - 1;

  while (low <= high) {
    const mid = Math.floor((low + high) / 2);
    const guess = searchList[mid];

    if (guess === value) {
      return mid;
    }

    if (guess < value) {
      low = mid + 1;
    } else {
      high = mid - 1;
    }
  }

  return null;
}

function main() {
  const searchList: number[] = [1, 2, 3, 4, 5, 6, 7, 8];
  const searchValue: number = 1;

  console.log("Binary Search");
  console.log("Search value: ", searchValue);

  const result = binarySearch(searchList, searchValue);
  const wrong = binarySearch(searchList, 12);

  console.log("Result: ", result);
  console.log("Wrong: ", wrong);
}

main();

File: 01_introduction_to_algorithms/go/binary-search.go
package main

import "fmt"

var (
	searchList = []int {1, 2, 3, 4, 5, 6, 7, 8}
	searchValue = 7
)

func BinarySearch(list []int, value int) int {
	var (
		low = 0
		high = len(list) - 1
	)

	for low <= high {
		var (
			mid = (low + high) / 2
			guess = list[mid]
		)

		if guess == value {
			return mid
		}

		if guess < value {
			low = mid + 1
		} else {
			high = mid - 1
		}
	}

	return -1
}

func main() {
	fmt.Println("Binary Search")

	fmt.Println("List: ", searchList)
	fmt.Println("Search value: ", searchValue)

	result := BinarySearch(searchList, searchValue)
	wrong := BinarySearch(searchList, 100000)

	fmt.Println("Result: ", result) // 6
	fmt.Println("Wrong: ", wrong) // -1
}
